B-Sockets-ServerTesting/sockets-servertesting.md

```diff
@@ -0,0 +1,287 @@
+---------------------------
+### What is a socket?
+
+A socket is a software construct used for many modes of communication between
+processes. The mode of communication that this recitation will focus on is
+network communication. In particular, stream sockets represent an endpoint for
+reliable, bidirectional connections such as TCP connections. This allows for two
+processes, on separate computers, to communicate over a TCP/IP network
+connection.
+
+Sockets have:
+- an IP address, to (typically) identify the computer that the socket endpoint
+  belongs to
+- a port number, to identify which process running on the computer the socket
+  endpoint belongs to
+- a protocol, such as TCP (reliable) or UDP (unreliable). Stream sockets use
+  TCP.
+
+An IP address and port number are both required in order for a computer to
+communicate with a specific process on a remote computer. Think about would
+happen if a port number wasn't necessary.
+
+### The client-server model
+The two endpoints in a socket connection serve different roles. One end acts as
+a *server*: 
+- It tells the operating system that it should recieve incoming connections on a
+  port number
+- It waits for incoming connections
+- When it recieves a connection, it creates a *new socket* for each client which
+  will be used for communication to that client
+
+The other end is a *client*:
+- It "connects" to the server using the server’s IP address and the port number
+
+After a client connects to a server, there is bidirectional communication
+between the two processes, often with I/O system calls such as `read()` and
+`write()`, or their socket-specific versions `recv()` and `send()`. 
+
+### Simple socket communication
+A simple way to demonstrate the bidirectional and network-based communcation of
+sockets is with `netcat`. `netcat` is a bare-bones program to send streams of
+binary data over the network.
+
+Imagine we have two computers that can communicate over the internet, with the
+IP addresses `clac.cs.columbia.edu` and `clorp.cs.nyu.edu`.
+
+Because of the client-server model, connecting two socket endpoints to each
+other is not a symmetrical process. One socket needs to act as the server, while
+the other needs to act as a client. You tell `netcat` to act as a server with
+the `-l` flag:
+```bash
+joy@clac.cs.columbia.edu:~$ nc -l 10000
+```
+
+The `netcat` program on `clac.cs.columbia.edu` will create a socket and wait for
+connections on the port 10000. To tell `netcat` to act as a client, you supply
+the IP address of the server and the port number of the socket listening on that
+server:
+```bash
+jeremy@clorp.cs.nyu.edu:~$ nc clac.cs.columbia.edu 10000
+```
+Notice the differences between these two commands. The first command only
+requires a port number, and doesn't need to know the IP address of the other
+computer. The second command requires knowledge of both the IP address -- what
+computer to connect to; and the port number -- which process to connect to on
```

Comment:
This formatting is weird... maybe change to "the IP address (which computer to connect to) and the port number (which process to connect to on that computer)"